From f5adc5e98d3531185dc5c1d5b6cc8262482701d2 Mon Sep 17 00:00:00 2001
From: Brian Bennett <brian.bennett@joyent.com>
Date: Wed, 13 Jun 2018 07:13:31 -0700
Subject: [PATCH] MANATEE-397 `manatee-adm history` fails with `TypeError:
 Cannot read property 'zoneId' of null` (again) Reviewed by: Josh Wilsdon
 <josh@wilsdon.ca> Reviewed by: Richard Bradley <richard.bradley@joyent.com>
 Approved by: Richard Bradley <richard.bradley@joyent.com>

---
 lib/adm.js | 13 ++++++++++++-
 1 file changed, 12 insertions(+), 1 deletion(-)

diff --git a/lib/adm.js b/lib/adm.js
index 92b2636..0b6de68 100644
--- a/lib/adm.js
+++ b/lib/adm.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2016, Joyent, Inc.
+ * Copyright (c) 2018, Joyent, Inc.
  */
 
 /**
@@ -1927,6 +1927,17 @@ function annotateHistoryNode(evt, last)
     }
 
     if (nst.generation > lst.generation) {
+        if (lst.sync === null && lst.oneNodeWriteMode !== true) {
+            /*
+             * We seem to have arrived here due to an invalid state on the
+             * previous generation, namely the previous generation is
+             * multi-mode, writable, with no replicas. This should never
+             * happen, but somehow it did. At any rate, we're just happy that
+             * there's a replica now.
+             */
+            return (sprintf('sync "%s" added',
+                nst.sync.zoneId.substr(0, 8)));
+        }
         if (nst.sync.zoneId == lst.sync.zoneId) {
             return ('error: gen number changed, but ' +
                 'primary and sync did not');
-- 
2.21.0

